File: 48_rotate_image/rotate.go
func rotate(matrix [][]int) {
	reverse(matrix)
    transpose(matrix)
}

func reverse(matrix [][]int) {
	l, r := 0, len(matrix)-1

	for l < r {
		matrix[l], matrix[r] = matrix[r], matrix[l]
		l++
		r--
	}
}

func transpose(matrix [][]int) {
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < i; j++ {
			matrix[i][j], matrix[j][i] =  matrix[j][i], matrix[i][j]
		}
	}
}

File: 48_rotate_image/rotate.py
from typing import List


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.reverse(matrix)
        self.transpose(matrix)
        
    def reverse(self, matrix: List[List[int]]) -> None:
        l, r = 0, len(matrix) - 1

        while l < r:
            matrix[l], matrix[r] = matrix[r], matrix[l]
            l += 1
            r -= 1

    def transpose(self, matrix: List[List[int]]) -> None:
        for i in range(len(matrix)):
            for j in range(i):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]